Improve the hooks provided by the visitor class

DerekStride opened the pull request:
## What

Adds support for `around_*` hooks that must yield to continue visiting child nodes. This allows implementing visitors that can stop traversing the tree.

It also allows running callback logic before & after visiting child nodes.

## Questions

The new `around_*` hooks make the `on_*` hooks unnecessary. Should we remove support for the `on_*` hooks? It would be a breaking change but we haven't released v1 yet so it could be a good time.

## Rendered Documentation Sample

<img width="679" alt="Screenshot 2023-03-30 at 2 42 34 PM" src="https://user-images.githubusercontent.com/6456191/228933718-c63cda33-e07f-4f13-9aab-7613182ca173.png">

DerekStride committed:
feat(hooks): Improve the hooks provided by the visitor class

Adds support for `around_*` hooks that must yield to continue visiting
child nodes. This allows implementing visitors that can stop traversing
the tree.

It also allows running callback logic before & after visiting child
nodes.

phsurette commented on lib/tree_stand/visitor.rb:
```diff
@@ -30,6 +35,31 @@ module TreeStand
+  #       @stack = []
+  #     end
+  #
+  #     def _around_default(node)
```
It seem strange to overrride `_around_default`. The leading underscore says to me this is an internal API which for some reason isn't private but should nevertheless be left alone.

DerekStride replied:
The reason I chose the leading underscore is to avoid conflicts if a grammar uses `default` as a node type. In that case there would be no way to define a hook just for that node e.g. `on_default`.

We can use change the API to something more intuitive like `default_on_hook` & `default_around_hook`.

phsurette replied:
I think those names are better enough to make the change.

DerekStride replied:
Alternatives:

- `default_around`, `default_on`
- `around`, `on` <- I'll try this
- `visit_around`, `visit_on`

phsurette reviewed (approved):
Looks good. I think it make sense to keep the `on_xxx` support, it's a little simpler to reason about.

These test cases seem a little complex to me. I would add a test case with an around visitor that doesn't always (or ever) yield to the block passed to it since that's a key benefit to the `around` paradigm.

DerekStride committed:
refactor(visitor)!: rename default visitor hook

BREAKING CHANGE: Renames the public method `_on_default` to `on` for the
TreeStand::Visitor class.

DerekStride commented:
I've update the PR according to the feedback:

- simplified the visitor tests
- added tests for when the `around` callback doesn't yield
- added some visual documentation to the visitor tests to help gain context on the shape of the tree
- Added a commit with a breaking change to rename the `_on_default` hook to `on`.
